### ISSUE TITLE
Advanced options: Custom triggers for effects.

### ISSUE THREAD

**Issue by cp2004** [cp2004/OctoPrint-WS281x_LED_Status#125]
**This ticket intends to track the sole idea of custom effects, which will be on the cards for 0.8.0**

**Please do not open a new issue requesting a specific event/trigger to be added to the plugin**

---

The predefined, hard coded events that the plugin reacts to clearly do not do *everything*. The proposed solution is user-configurable 'triggers', that can be used to customise any effects.#102 

**What will this involve?**
* [x] Defining the scope and limits of this, what will be available
* [x] Defining a structure of settings and relationships between trigger and effect
* [x] Implementing the various event subscriptions in the backend
* [x] Implementing a frontend to configure the effects

**What will *not* be covered by this issue**
* Custom progress bars
* Per segment control of LEDs
* Removing the predefined setups. They may be converted to use the same 'infrastructure' but will not be removed.

---

Listing registered events

```python
events = [x for x in octoprint.events.Events.__dict__ if x.isupper()]
```

--- 

## Custom Triggers

* User defined @ command - `@WS CUSTOM <command>` - need to distinguish between custom vs. reserved for standard ones
* Any OctoPrint event - this would include all those registered by other plugins in addition to those on https://docs.octoprint.org
* Any gcode command - G/M code, exact match, regex (?)
  * Need to think about performance here, could easily build up enough searches for every command that we slow down the comm layer. Maybe add all lines to a queue and asynchronously process them? Can work out ahead of time if there are any searches that need to be done, so if it is a performance issue it can be disabled.

## Settings Structure

```json
{
    "custom":[
        {
            "type": "atcommand",
            "command": "party",
            "effect": "Solid Colour",
            "color": "#ff0000",
            "delay": "1"
        },
        {
            "type": "gcode",
            "match_type": "gcode/exact/regex",
            "pattern": "G28",
            "effect": "Solid Colour",
            "color": "#ff0000",
            "delay": "1"
        },
        {
            "type": "event",
            "event": "PrintDone",
            "effect": "Solid Colour",
            "color": "#ff0000",
            "delay": "1"
        }
    ]
    
}
```

**Comment by cp2004** [comment]
Custom triggers are go! Implemented ready for 0.8.0.

![image](https://user-images.githubusercontent.com/31997505/142676518-138cea41-e290-41c5-a934-d509aa331435.png)

**Comment by cp2004** [comment]
0.8.0 has been released (finally...!)
